Strata/Languages/Laurel/Grammar/ConcreteToAbstractTreeTranslator.lean

```diff
@@ -0,0 +1,181 @@
+/-
+  Copyright Strata Contributors
+
+  SPDX-License-Identifier: Apache-2.0 OR MIT
+-/
+
+import Strata.DDM.AST
+import Strata.Languages.Laurel.Grammar.LaurelGrammar
+import Strata.Languages.Laurel.Laurel
+import Strata.DL.Imperative.MetaData
+import Strata.Languages.Boogie.Expressions
+
+namespace Laurel
+
+open Laurel
+open Std (ToFormat Format format)
+open Strata (QualifiedIdent Arg SourceRange)
+open Lean.Parser (InputContext)
+open Imperative (MetaData Uri FileRange)
+
+structure TransState where
+  inputCtx : InputContext
+  errors : Array String
+
+abbrev TransM := StateM TransState
+
+def TransM.run (ictx : InputContext) (m : TransM α) : (α × Array String) :=
+  let (v, s) := StateT.run m { inputCtx := ictx, errors := #[] }
+  (v, s.errors)
+
+def TransM.error [Inhabited α] (msg : String) : TransM α := do
+  modify fun s => { s with errors := s.errors.push msg }
+  return panic msg
+
+def SourceRange.toMetaData (ictx : InputContext) (sr : SourceRange) : Imperative.MetaData Boogie.Expression :=
+  let file := ictx.fileName
+  let startPos := ictx.fileMap.toPosition sr.start
+  let endPos := ictx.fileMap.toPosition sr.stop
+  let uri : Uri := .file file
+  let fileRangeElt := ⟨ Imperative.MetaDataElem.Field.label "fileRange", .fileRange ⟨ uri, startPos, endPos ⟩ ⟩
+  #[fileRangeElt]
+
+def getArgMetaData (arg : Arg) : TransM (Imperative.MetaData Boogie.Expression) :=
+  return arg.ann.toMetaData (← get).inputCtx
+
+def checkOp (op : Strata.Operation) (name : QualifiedIdent) (argc : Nat) :
+  TransM Unit := do
+  if op.name != name then
+    TransM.error s!"Op name mismatch! \n\
+                   Name: {repr name}\n\
+                   Op: {repr op}"
+  if op.args.size != argc then
+    TransM.error s!"Op arg count mismatch! \n\
+                   Expected: {argc}\n\
+                   Got: {op.args.size}\n\
+                   Op: {repr op}"
+  return ()
+
+def translateIdent (arg : Arg) : TransM Identifier := do
```

Comment:
I think using the #strata_gen representation will lead to cleaner code and avoid the need for all the qualified expression references.